### PR TITLE
Fix bundle. Up for discussion :)

### PR DESCRIPTION
**Pull request by Baune8D**
So i just started using UnderStrap which is excellent, but i made some changes to this child-theme to make it easier to work with.

This is not a fix for something specific but a collection of fixes, you dont neccesarily need to merge this, but there might be some interesting stuff in it.

So this is the following changes:

* Move package.json build dependencies to devDependencies section.
* Add local Gulp so project is not dependent on a global install.
* Add NPM scripts which will make it easier to start development:
  * init - Do npm install, gulp copy-assets and initial build in one command.
  * watch - Use local Gulp to start watch.
  * watch-bs - Use local Gulp to start watch-bs.
  * start - for easy start of watch with just: npm start
* Correct package.json formatting and add version constraints to packages.
* Update NPM packages and add package-lock.json to avoid dependency errors.
* Compile scripts and styles on watch start.
* Better support for using this template in another git repository:
  * Git ignore all gulp output files to avoid constant changes to repo. (This will force you to at least build the project once after clone)
  * Align variable with parent in sass/theme/_child_theme_variables.scss so we wont see git change after initial build.
* Add VS Code tasks for easier use in that IDE.
* Adjust README for all the mentioned changes.

### CODE REVIEW

**Comment by IngoVals** [comment]
This sounds great, havent had chance to code review though.

> Git ignore all gulp output files to avoid constant changes to repo. (This will force you to at least build the project once after clone)

Note that this would make it unusable straight out of WordPress install but imo a child theme shouldn't be used like that.

**Comment by Baune8D** [comment]
@IngoVals Yes i know :) 

For my use-case, i have a Github repository setup for my modified child theme, and seeing changes to the repository every time it does a build can be quite annoying.

Since no one would really use a child theme without any modifications, i think it should be less of a problem.

**Comment by Baune8D** [comment]
@IngoVals In the modified README i mention that you have to do an initial "npm run init" which will do everything necessary to make it work.

**Comment by IngoVals** [comment]
@holger1411 @axlright Check this out.

How we want to approach this depends mostly on how we see the child theme being used.

**Comment by axlright** [comment]
Thanks @IngoVals - I agree that a larger (much larger?) conversation about the child theme needs to happen. The current setup involves both child and parent needing to be updated, and that confuses many (all?) people.

I'm not sure where to have the conversation. A Pull Request comment thread is probably not the place! @holger1411 do you have a suggestion for a place to discuss the relationship between parent and child?

My understanding is that the parent (of course) contains everything - all template and function files as well as all Bootstrap and Underscores as well as the gulp / npm configuration.

Then the child contains none (very few) of the templates and functions and instead references the parent as a dependency. This is standard. 

Where it gets weird is that the child also has all Bootstrap and Underscores as well as its own gulp / npm configuration. Styling happens in the child theme but references Sass files locally. Same with scripts.

My understanding for the reasoning for this is because people who don't have npm / gulp should be able to use the child theme - the base styles are all there for them to add on to (?) - I could be wrong with this. Maybe it's also because of requirements for inclusion in the WordPress theme directory.

I'd personally like to see a day where the parent has **everything** and is the only thing that gets updated. Then, npm / gulp happens in _either_ the parent _or_ the child. The only thing different is the gulp paths for the child are set with a bunch of `../../` type paths to look to the parent as a dependency for everything - scss / js and all npm dependencies including all of Boostrap and Underscores and Font Awesome. 

Then if you develop via npm / gulp in the child, it writes out the final app.js and style.css (or whatever people call them) in the child theme, but if you develop via npm / gulp in the parent, those files are written out in the parent. And the functions.php for each calls appropriately.

If you wanted to load custom dependencies, you could do so equally well in the parent or in the child as there would be paths appropriate for each.

Then of course only the parent needs to be updated and if a developer has copied a file from the parent to the child, she's responsible - like always - for keeping it updated.

The workflow now is more complicated.

I'd love to figure out a place to discuss this.

Ideas?

**Comment by holger1411** [comment]
@axlright 
Your child theme understanding is basically right for a normal WordPress child theme.
But the UnderStrap child theme uses its own dependencies like Bootstrap etc. Just the .php files are used from the parent theme.
By default, both, parent and child, using the same dependencies. So the result will be the same.

But if you need to change the dependencies you can do it indipendently from the parent and you can rebuild your css and js distribution without touching the parent theme (and without loosing the updatebility).
So for example if there is a new Bootstrap version you can change it in your child theme and rebuild your sources and run your website with this new version without editing one line in the parent theme. Even if there are changes in .php files you can copy them from your parent to your child and rewrite the markup if necessary.

So you have two independent sets of dependencies.

It also means that your website still works if you update the parent theme (which, for example again, uses a newer Bootstrap version), because your child theme still uses the older version.
So an update will not break live websites.
It just means you have to update your child themes dependencies to be on the latest version.

**Comment by axlright** [comment]
@holger1411 I hear you and defer to you as you're the main man! I still feel that the current workflow creates some other issues along the way. 

If I understand your explanation, the reason that the child and parent both contain Bootstrap dependencies is so that you can update the dependencies of the child theme without updating the php of the parent theme? So the use case here is if you don't want to get php conflicts from a parent theme update but you do want to update the dependencies?

To me personally, this seems like an edge case. But it may just be me? How often does this above circumstance happen?

The cost is that both the parent and the child need to be updated - the parent for the php and the child for the Bootstrap an other dependencies. Moving to a system where the child theme looked to the parent theme for all dependencies would mean you'd never have to update the child theme, though you'd always have to get php and dependency updates together.

I'm curious if others in the community have thoughts about this. I don't know where the best place to discuss it is. I do hear from new users that they are confused, so if we keep the current workflow, the documentation should better describe the process.

Thanks!

**Comment by jacob87o2** [comment]
This is nice. 🙂  Relates to the question I asked here: #239 

What do you think about ignoring the `src` folder also?

In my opinion git repos should be used to track source files and output should be kept in separate artifact repository. But again - that's just my opinion. Many projects do include `dist` folder in its git repo... but it's at least separate folder with tree-shaken dependencies. And still - not something I would personally do.

File: package.json [hunk 1]
```diff
@@ -35,6 +39,7 @@
     "browser-sync": "^2.23.6",
     "del": "^3.0.0",
     "gulp": "3.9.1",
+    "gulp-cli": "^2.0.1",
```
Comment (jacob87o2) [inline package.json]: I think, the `gulp-cli` is just a tool to run gulp, and it's supposed to be installed globally. 

The `gulp` module is the one that will do the work and it's already included in the package.json file.

File: package.json [hunk 2]
```diff
@@ -4,7 +4,11 @@
   "description": "Basic Child Theme for UnderStrap Theme Framework: https://github.com/holger1411/understrap",
   "main": "index.js",
   "scripts": {
-    "postinstall": "gulp copy-assets"
+    "postinstall": "gulp copy-assets",
+    "init": "npm install && npm run postinstall",
```
Comment (jacob87o2) [inline package.json]: npm already runs `postinstall` after npm install

https://docs.npmjs.com/misc/scripts

**Pull request closed without merging**